File: [0300][Medium][Longest_Increasing_Subsequence]/Longest_Increasing_Subsequence_2.py
# leetcode time     cost : 52 ms
# leetcode memory   cost : 13.9 MB 
# Time  Complexity: O(N*logN)
# Space Complexity: O(N)
# slolution 4, DP and binarySearch with bisect lib,
import bisect
class Solution:
    #def lengthOfLIS(self, nums: List[int]) -> int:
    def lengthOfLIS(self, nums):
        # minend[i] is the minimum ending of an increasing subsequence of length i+1.
        minend = [float('inf')] * (len(nums) + 1)
        for num in nums:
            minend[bisect.bisect_left(minend, num)] = num
            # at last, minend is [1, 3, 4, 5, 6, 10, inf, inf, inf, inf]
        # to find the first elment of float('inf') in the minend[], just the max length index     
        return minend.index(float('inf'))
    
    def lengthOfLIS_2(self, nums):
        minend = [float('inf')] * len(nums)
        for num in nums:
            minend[bisect.bisect_left(minend, num)] = num
        return bisect.bisect_left(minend, float('inf'))

def main():
    nums = [1,3,6,7,9,4,10,5,6] #expect is 6
    obj = Solution()
    result = obj.lengthOfLIS(nums)
    print("return result is "+str(result));

if __name__ =='__main__':
    main() 

File: [0300][Medium][Longest_Increasing_Subsequence]/Longest_Increasing_Subsequence_2.java
// leetcode time     cost : 173 ms
// leetcode memory   cost : 101.2 MB 
// Time  Complexity: O(n**2)
// Space Complexity: O(n**2)
// slolution 2, recursion with memory,
public class Solution {
    public int lengthOfLIS(int[] nums) {
        int memo[][] = new int[nums.length + 1][nums.length];
        for (int[] l : memo) {
            Arrays.fill(l, -1);
        }
        return lengthofLIS(nums, -1, 0, memo);
    }
    public int lengthofLIS(int[] nums, int previndex, int curpos, int[][] memo) {
        if (curpos == nums.length) {
            return 0;
        }
        
        // directly use the value if memo[previndex + 1][curpos] is calculated before,
        // memo[previndex + 1][curpos] is the max window length 
        // if num[curpos] chosen and second end in window is nums[previndex + 1] when it comes to curpos, 
        if (memo[previndex + 1][curpos] >= 0) {
            return memo[previndex + 1][curpos];
        }
        // for every nums[currpos], the max window come to this position is memo[previndex + 1]=window[]
        // the winidow will be passed to the deeper level for recursion.
        int taken = 0;
        if (previndex < 0 || nums[curpos] > nums[previndex]) {
            taken = 1 + lengthofLIS(nums, curpos, curpos + 1, memo);
        }

        int nottaken = lengthofLIS(nums, previndex, curpos + 1, memo);
        // compare to DP the [curpos] is the outside iterator i, [previndex + 1] is the inside iterator j(0<j<i).
        memo[previndex + 1][curpos] = Math.max(taken, nottaken);
        return memo[previndex + 1][curpos];
    }
}

public class Longest_Increasing_Subsequence_2 {
    public static void main(String args[]) {
        int[] nums = {1,3,6,7,9,4,10,5,6}  ;   // expect is 6
        Solution Solution_obj = new Solution();
        int result = Solution_obj.lengthOfLIS(nums);
        System.out.println("In java code,return value is :" + (result));
    }
}